src/lib/bark/services/user-account-service.ts

```diff
@@ -0,0 +1,153 @@
+import { Err, Ok, Result } from "@/lib/utilities/result";
+import { BarkContext } from "../bark-context";
+import {
+  UserAccount,
+  UserAccountSpec,
+  UserAccountSpecSchema,
+  UserAccountUpdate,
+  UserIdentifier,
+} from "../models/user-models";
+import { CODE } from "@/lib/utilities/bark-code";
+import { dbBegin, dbCommit, dbRelease, dbRollback } from "@/lib/data/db-utils";
+import { EncryptedUserAccountDao } from "../daos/encrypted-user-account-dao";
+import {
+  toEncryptedUserAccountSpec,
+  toUserAccount,
+} from "../mappers/user-mappers";
+import { LRUCache } from "lru-cache";
+
+export class UserAccountService {
+  private idCache: LRUCache<string, UserIdentifier>;
+
+  constructor(private context: BarkContext) {
+    this.idCache = new LRUCache({ max: 10 });
+  }
+
+  async create(args: {
+    spec: UserAccountSpec;
+  }): Promise<
+    Result<
+      UserIdentifier,
+      typeof CODE.FAILED | typeof CODE.ERROR_ACCOUNT_ALREADY_EXISTS
+    >
+  > {
+    const { spec } = args;
+    const conn = await this.context.dbPool.connect();
+    try {
+      await dbBegin(conn);
+      const encrypted = await toEncryptedUserAccountSpec(this.context, spec);
+      const { userHashedEmail } = encrypted;
+      const dao = new EncryptedUserAccountDao(conn);
+      const existing = await dao.getByUserHashedEmail({ userHashedEmail });
+      if (existing !== null) {
+        return Err(CODE.ERROR_ACCOUNT_ALREADY_EXISTS);
+      }
+      const identifier = await dao.insert({ spec: encrypted });
+      await dbCommit(conn);
+      return Ok(identifier);
+    } catch (err) {
+      console.error(err);
+      return Err(CODE.FAILED);
+    } finally {
+      await dbRelease(conn);
+    }
+  }
+
+  async applyUpdate(args: {
+    userId: string;
+    update: UserAccountUpdate;
+  }): Promise<
+    typeof CODE.OK | typeof CODE.FAILED | typeof CODE.ERROR_USER_NOT_FOUND
+  > {
+    const { context } = this;
+    const { dbPool } = context;
+    const { userId, update } = args;
+    const conn = await dbPool.connect();
+    try {
+      await dbBegin(conn);
+      const dao = new EncryptedUserAccountDao(conn);
+      const encryptedAccount = await dao.getByUserId({ userId });
+      if (encryptedAccount === null) {
+        return CODE.ERROR_USER_NOT_FOUND;
+      }
+      const account = await toUserAccount(context, encryptedAccount);
+      const merged = { ...account, ...update };
+      const { userEmail, userTitle, userName, userPhoneNumber, userResidency } =
+        merged;
+      const spec: UserAccountSpec = {
+        userEmail,
+        userTitle,
+        userName,
+        userPhoneNumber,
+        userResidency,
+      };
```

Comment:
Is it possible to use User Account Spec Schema to parse merged such that it returns just the fields of a spec?

Reply:
Yes it is possible! See 10af4e16

Reply:
Updated this instance in 5aac5924 - will find other cases as we go.